### PR TITLE
Add support for passing in custom environment variables

### PR DESCRIPTION
**Pull request by mglazer**
Includes extremely limited support for some very limited
template expansion, currently limited to just expanding
the current working directory of the process.

---

This change is [Reviewable](https://reviewable.io/reviews/palantir/go-java-launcher/9)

### CODE REVIEW

**Comment by mglazer** [comment]
@uschi2000

File: README.md [hunk 1]
```diff
@@ -61,5 +68,10 @@ and `<custom.xyz>` refer to the options from the two configuration files, respec
 Note that the custom `jvmOpts` appear after the static `jvmOpts` and thus typically take precendence; the exact
 behaviour may depend on the Java distribution.
 
+`env` supports a very limited set of automatic expansions if you surround the expansion variable with `%%` as shown above,
```
Comment (uschi2000) [inline README.md]: suggestion:

```
The `env` block (both in static and custom configuration) supports a fixed set of automatic expansions for values assigned to environment variables:
...
```

File: README.md [hunk 2]
```diff
@@ -61,5 +68,10 @@ and `<custom.xyz>` refer to the options from the two configuration files, respec
...
+for `CUSTOM_PATH`. The following expansions are supported:
+
+* `%%.CWD%%`: The current working directory of the user which executed this process
```
Comment (uschi2000) [inline README.md]: low conviction: is `{{CWD}}` more canonical syntax?

File: launchlib/config_test.go [hunk 3]
```diff
@@ -68,4 +101,34 @@ jvmOpts:
...
+	if config.Env != nil {
+		t.Errorf("Expected environment to be nil, but was %v", config.Env)
+	}
+}
+
+func TestParseCustomConfigWithEnvPlaceholder(t *testing.T) {
+	var data = []byte(`
+configType: java
+configVersion: 1
+env:
```
Comment (uschi2000) [inline launchlib/config_test.go:114]: maybe add test asserting that

```
env:
  %%.CWD%%: foo
```

leads to failure

File: launchlib/config_test.go [hunk 4]
```diff
@@ -41,9 +44,13 @@ args:
 		ConfigVersion: 1,
 		MainClass:     "mainClass",
 		JavaHome:      "javaHome",
-		Classpath:     []string{"classpath1", "classpath2"},
-		JvmOpts:       []string{"jvmOpt1", "jvmOpt2"},
-		Args:          []string{"arg1", "arg2"}}
+		Env: map[string]string{
+			"SOME_ENV_VAR":  "/etc/profile",
+			"OTHER_ENV_VAR": "/etc/redhat-release",
```
Comment (uschi2000) [inline launchlib/config_test.go:49]: should test parsing behaviour for non-supported placeholder. in particular does `"%%.FOO%%BAR"` resolve to `""` or to `"%%.FOO%%BAR"`, or does it yield an exception? i'd favour the latter.

File: launchlib/config_test.go (hunk 3 above)
Comment (uschi2000) [reply launchlib/config_test.go:114]: (not really sure what the right place is for this test)

File: launchlib/config_test.go (hunk 4 above)
Comment (uschi2000) [reply launchlib/config_test.go:49]: (not really sure what the right place is for this test)

**Review by uschi2000 — CHANGES_REQUESTED** [review]
generally OK, but would like to see a little more test&doc clarification around edge cases, see detailed comments

File: README.md (hunk 2 above)
Comment (mglazer) [reply README.md]: Canonical given what reference world? 

{{}} feels weird to me because {} is actually a reserved character in YAML:

http://www.yaml.org/spec/1.2/spec.html

% is too, but it's used to denote a directive line. 

I wish I could get rid of the `.CWD`, but unsure how with Go templates.

File: launchlib/config_test.go (hunk 4 above)
Comment (mglazer) [reply launchlib/config_test.go:49]: your last comment is in reference to asking about adding variable expansion tests to this specific test (since that wouldn't make sense here given variable expansion happens at execution time)

File: launchlib/config_test.go (hunk 3 above)
Comment (mglazer) [reply launchlib/config_test.go:114]: So that would lead to failure because you didn't surround the %%  with quotes. So your YAML file is actually invalid, and will fail to parse at this point. Which I guess is the failure you were hoping for, but maybe not in the same way.

**Comment by mglazer** [comment]
Updated commit. I'll squash when this is all good, but for now, multiple commits so it's easier to diff.

**Comment by uschi2000** [comment]
Reviewed 2 of 5 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[README.md, line 74 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFp-r1-74:-KUeehFx-OMTFYjJvlwc:b7m11v6) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/README.md#L74)):_
<details><summary>

<i>Previously, mglazer wrote…</i></summary>



> Canonical given what reference world? 
> 
> {{}} feels weird to me because {} is actually a reserved character in YAML:
> 
> http://www.yaml.org/spec/1.2/spec.html
> 
> % is too, but it's used to denote a directive line. 
> 
> I wish I could get rid of the `.CWD`, but unsure how with Go templates.
> </details>
> OK, reading up on Go templates, it seems that they are powerful enough to call functions. That's not OK.

Regarding those markers, I'd prefer `BAR: "{{FOO}}"` over `BAR: %%.FOO%%` and in strings the `{,}` symbols won't be reserved.

---

_[launchlib/config_test.go, line 114 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFr-r1-114:-KUefdm7RDsNqAKDNsrH:bqruezn) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/launchlib/config_test.go#L114)):_
<details><summary>

<i>Previously, mglazer wrote…</i></summary>



> So that would lead to failure because you didn't surround the %%  with quotes. So your YAML file is actually invalid, and will fail to parse at this point. Which I guess is the failure you were hoping for, but maybe not in the same way.
> </details>
> ok, I meant with quotes. variable expansion is only allowed in values, not in fields.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by mglazer** [comment]
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

_[README.md, line 74 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFp-r1-74:-KUf-XyA1M_-YGYIlxpx:b-gx91u3) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/README.md#L74)):_
<details><summary>

<i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>



> OK, reading up on Go templates, it seems that they are powerful enough to call functions. That's not OK.
> 
> Regarding those markers, I'd prefer `BAR: "{{FOO}}"` over `BAR: %%.FOO%%` and in strings the `{,}` symbols won't be reserved.
> </details>
> Changed to {{}}. go templates aren't being used anymore (then again, I'm still unsure what the attack vector is. I would be really curious in hearing about it)

---

_[launchlib/config_test.go, line 114 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFr-r1-114:-KUf-Oz6OXexszGZ3jx7:b-qz82sh) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/launchlib/config_test.go#L114)):_
<details><summary>

<i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>



> ok, I meant with quotes. variable expansion is only allowed in values, not in fields.
> </details>
> Are the new set of tests sufficient to cover this?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by mglazer** [comment]
@uschi2000 : Any further changes you'd like me to make here?

**Comment by uschi2000** [comment]
yep, one nitpick on the template method, otherwise this is fine

---

Review status: 1 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

_[README.md, line 74 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFp-r1-74:-KUsmEntJlqfcLND7lWp:b-k47ecw) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/README.md#L74)):_
<details><summary>

<i>Previously, mglazer (Mike Glazer) wrote…</i></summary>



> Changed to {{}}. go templates aren't being used anymore (then again, I'm still unsure what the attack vector is. I would be really curious in hearing about it)
> </details>
> the attack vector is only with Go Templates. since you switched to string.Replacer, this is fine now.

---

_[launchlib/config_test.go, line 114 at r1](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUed--YKfX1u1NrLDFr-r1-114:-KUsmC-yLJP0bjldquWC:b-rx3qis) ([raw file](https://github.com/palantir/go-java-launcher/blob/997357481a9fce6c935b0bec28d53e75d77df08e/launchlib/config_test.go#L114)):_
<details><summary>

<i>Previously, mglazer (Mike Glazer) wrote…</i></summary>



> Are the new set of tests sufficient to cover this?
> </details>
> yep, thanks

---

_[launchlib/launcher.go, line 121 at r3](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUslPgAUyNlVu6iQMF7:-KUslPgBdqLkrQE9ReKx:bz6tys8) ([raw file](https://github.com/palantir/go-java-launcher/blob/58446acac2b387d49b5605957637ca5d90eaed93/launchlib/launcher.go#L121)):_

> ``` Go
> }
> 
> func fillEnvironmentVariables(env map[string]string, customEnv map[string]string) map[string]string {
> ```
- non-trivial method, needs comment
- this method does two things: merge maps and replace `{{FOO}}` strings. can they be two functions? e.g., `env = expandTemplates(merge(staticConfig.env, customConfig.env))`
- decide whether you want to return the map or mutate it in place; it currently does both

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-:-KUsmNgJs4LZP-JuMzGg:boev52w)_

**Comment by uschi2000** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by mglazer** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[launchlib/launcher.go, line 121 at r3](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUslPgAUyNlVu6iQMF7:-KUt4qdHkIn7dwuyjxvb:bb2fabl) ([raw file](https://github.com/palantir/go-java-launcher/blob/58446acac2b387d49b5605957637ca5d90eaed93/launchlib/launcher.go#L121)):_
<details><summary>

<i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>



> - non-trivial method, needs comment
> - this method does two things: merge maps and replace `{{FOO}}` strings. can they be two functions? e.g., `env = expandTemplates(merge(staticConfig.env, customConfig.env))`
> - decide whether you want to return the map or mutate it in place; it currently does both
> 
> </details>
> Is there anything wrong with doing both?  Or just from a fundamental standpoint you'd prefer that it's either pure or impure, not both?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by uschi2000** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[launchlib/launcher.go, line 121 at r3](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-KUslPgAUyNlVu6iQMF7:-KUt7Qz4nRZ_Lj54LYWs:b-bxgdri) ([raw file](https://github.com/palantir/go-java-launcher/blob/58446acac2b387d49b5605957637ca5d90eaed93/launchlib/launcher.go#L121)):_
<details><summary>

<i>Previously, mglazer (Mike Glazer) wrote…</i></summary>



> Is there anything wrong with doing both?  Or just from a fundamental standpoint you'd prefer that it's either pure or impure, not both?
> </details>
> it's just easier to read, that's all

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by uschi2000** [comment]
Reviewed 1 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9)_

**Comment by uschi2000** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/go-java-launcher/9#-:-KUtAM1MaoYusrXyeb7A:bnfp4nl)_

**Comment by uschi2000** [comment]
Cheers Mike!
